internal/provider/provider.go

```diff
@@ -0,0 +1,253 @@
+package provider
```

Comment:
If you'd like to review this file with more incremental changes, they're reasonably well broken down by commit:

* The first commit in this PR lifts code from the old `provider.go` file without much change except to pull bits out into other packages.
* The second commit switches to using the Vault API package instead of a raw HTTP client
* The third commit ensures that `ctx` values do not go ignored in those Vault API calls